tests/pytests/functional/cache/test_cache.py

```diff
@@ -0,0 +1,526 @@
+    cache = salt.cache.factory(opts)
+    yield cache
+
+
+@pytest.fixture
+def localfs_cache(minion_opts):
+    opts = minion_opts.copy()
+    opts["cache"] = "localfs"
+    cache = salt.cache.factory(opts)
+    yield cache
+    try:
+        shutil.rmtree(opts["cachedir"])
+    except Exception:  # pylint: disable=broad-except
+        # Not my circus, not my monkey
+        pass
+
+
+@pytest.fixture
+def consul_cache(minion_opts, consul_port, consul_container):
+    opts = minion_opts.copy()
+    opts["cache"] = "consul"
+    opts["consul.host"] = "127.0.0.1"
+    opts["consul.port"] = consul_port
+    # NOTE: If you would like to ensure that alternate suffixes are properly
+    # tested, simply change this value and re-run the tests.
+    opts["consul.timestamp_suffix"] = ".frobnosticate"
+    cache = salt.cache.factory(opts)
+    yield cache
+
+
+@pytest.fixture
+def mysql_cache(minion_opts, mysql_port, mysql_container):
+    # We're doing a late import because we need access to the exception
+    import salt.cache.mysql_cache
+
+    # The container can be available before mysql actually is
+    mysql_container.container.exec_run(
+        ["/bin/sh", "-c", 'while ! mysql -e "SELECT 1;" >/dev/null; do sleep 1; done'],
+    )
+
+    # Gotta make the db we're going to use
+    res = mysql_container.container.exec_run(
+        ["/bin/sh", "-c", 'echo "create database salt_cache;" | mysql'],
+    )
+
+    opts = minion_opts.copy()
+    opts["cache"] = "mysql"
+    opts["mysql.host"] = "127.0.0.1"
+    opts["mysql.port"] = mysql_port
+    opts["mysql.user"] = "root"
+    opts["mysql.database"] = "salt_cache"
+    opts["mysql.table_name"] = "cache"
+    cache = salt.cache.factory(opts)
+
+    # For some reason even though mysql is available in the container, we
+    # can't reliably connect outside the container. Wait for access
+    timer = Timer(timeout=10)
+    while not timer.expired:
+        try:
+            # Doesn't matter what. We just have to execute so that we spin
+            # here until we can actually connect to the db instance.
+            cache.modules["mysql.list"]("fnord")
+        except salt.cache.mysql_cache.MySQLdb.DatabaseError:
+            # We don't really care what MySQL error is happening -
+            pass
+        else:
+            break
+    else:
+        assert False, 'Timer expired before "select 1;" worked'
+
+    # This ensures that we will correctly alter any existing mysql tables for
+    # current mysql cache users. Without completely altering the mysql_cache
+    # implementation there's no real other reasonable way to reset the client
+    # and force the alter_table to be called. Resetting the client to `None` is
+    # what triggers the implementation to allow the ALTER TABLE to add the
+    # last_update column
+    run_query = cache.modules["mysql.run_query"]
+    run_query(
+        conn=None,
+        query="ALTER TABLE salt_cache.cache DROP COLUMN last_update",
+    )[0].fetchone()
```

Comment:
Mark as destructive tests?
But isn't mysql running in a container here?

Reply:
Well, it shouldn't need to be destructive, because containers.

The problem that this is testing is that *existing* cache users won't have the column on the table. And since when we first connect we're going to create the table *with* the column, I needed a way to test that we'd also alter existing tables.

Which this will do (and is called out in the changelog)

Reply:
But if the test runs in a container, we start, we're not destroying the system the test suite is running on...